torch/csrc/jit/script/compiler.cpp

```diff
@@ -1424,24 +1421,29 @@ struct to_ir {
 
   // Desugars slice syntactic sugar tensor[begin:end] -> tensor.slice(begin,
   // end).
-  Value* emitSlice(
-      const SourceRange& loc,
-      TreeList&& inputs) {
-    const auto applyInputs =
-        Compound::create(TK_LIST, loc, std::move(inputs));
-    const auto input_values = getNamedValues(applyInputs->trees(),
-                                             /*maybe_unpack*/false,
-                                             identity);
+  Value* emitSlice(const Slice& slice) {
+    const auto& loc = slice.range();
+    TreeList inputs = {slice.value(), slice.startOr(0)};
+    const auto applyInputs = Compound::create(TK_LIST, loc, std::move(inputs));
+    const auto input_values = getNamedValues(
+        applyInputs->trees(),
+        /*maybe_unpack*/ false,
+        identity);
+
     NamedValue tensor = input_values[0];
     NamedValue begin = input_values[1];
-    NamedValue end = input_values[2];
-    NamedValue dim = NamedValue(loc, "dim",
-        graph->insertConstant(0, loc));
-    NamedValue step = NamedValue(loc, "step",
-        graph->insertConstant(1, loc));
+    NamedValue dim = NamedValue(loc, "dim", graph->insertConstant(0, loc));
+    NamedValue step = NamedValue(loc, "step", graph->insertConstant(1, loc));
+
+    std::vector<NamedValue> args = {tensor, dim, begin};
+    const auto has_end = slice.end().present();
+    if (has_end) {
+      // If the user specified an `end` index, pass it down
+      args.emplace_back(loc, "end", emitExpr(Expr(slice.end().get()), identity));
+    }
```

Comment:
Start is optional too

Reply:
Using `slice.startOr(0)` works because the beginning of the range is inclusive. If you'd like, I can make `start` explicitly optional for the purpose of schema matching, but I thought it might be easier to read to avoid the four-way branching.

Reply:
Oh, right I missed that and was confused by the asymmetry here. Let's leave it as is, thanks!